platform/mds/mds/src/main/java/org/motechproject/mds/service/impl/EntityServiceImpl.java

```diff
@@ -964,9 +1014,68 @@ public int compare(Field o1, Field o2) {
             fieldDtos.add(field.toDto());
         }
 
+        if(withPseudofields) {
+            EncodePseudofields(entity, fieldDtos);
+        }
+
         return addNonPersistentFieldsData(fieldDtos, entity);
     }
 
+    /***
+     * Adds display-only fields to the field list so that we can reuse UI widgets and
+     * wire protocol for editing entities with @Discriminated.
+     *
+     * One field is added for the discriminator.
+     * Once a discriminator is selected, the UI will expand the field list using fieldsAdded.
+     *
+     * Decode performed by:
+     *     org.motechproject.mds.web.service.impl.InstanceServiceImpl#DecodePseudofields
+     */
+    private void EncodePseudofields(Entity entity, List<FieldDto> fieldDtos) {
+        if(entity.isDiscriminated()) {
+            List<String> values = new ArrayList<String>();
+            List<EntityDto> entitiesDerived = new ArrayList<>();
+            int c=0;
+            List<EntityDto> deriveds = this.findEntitiesBySuperclass(entity.getClassName());
+            if(deriveds != null) {
+                for (EntityDto derived : deriveds) {
+                    values.add(derived.getName());
+                    entitiesDerived.add(derived);
+                    c++;
+                }
+            }
+            Set<String> fieldids = new HashSet<String>();
+            for(FieldDto fd : fieldDtos) {
+                fieldids.add(fd.getBasic().getName());
+            }
+
+            for (EntityDto e : entitiesDerived) {
+                List<FieldDto> fieldsAdded = new LinkedList<FieldDto>();
+                for (FieldDto fd2 : this.getFields(e.getId())) {
+                    if(!fieldids.contains(fd2.getBasic().getName()))
+                        fieldsAdded.add(fd2);
+                }
+                e.setFieldsAdded(fieldsAdded);
+            }
+
+            Type tD = allTypes.retrieveByClassName("java.util.Collection");
```

Comment:
and possibly other ways of clearing things up, I would see this logic getting split into smaller methods.